First version of Durable Function chaining lesson

lechnerc77 opened the pull request:
Referencing issue #2 
I have added a first version of the Durable Functions lesson for function chaining

* the main content of the lesson is available in the directory `\lessons\durablefunctions\`
* the sources code for the lesson are available in the directory `\src\durablefunctions\chaining\ts`
   * the code for the main part is available in the directory `.\DurableFunctionApp`
   * the code for the retry setup is available in the directory `.\DurableFunctionApp-Retry`
   * the code for the timeout setup is available in the directory `DurableFunctionApp-Timeout`
* the homework instruction is available in the directory `\homework\durablefunctions\chaining`
   * the solution of the homework is available in the directory `\src\homework\durablefunctions\chaining\ts`

As stated in the issue, the description and the code is focusing on TypeScript

marcduiker commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,631 @@
+|1.2| [The Orchestrator Function](#12-the-orchestrator-function)
+|1.3| [The Activity Function](#13-the-activity-function)
+|1.4| [The First Execution](#14-the-first-execution)
+|2| [Implementing a "Real-World" Scenario](#2-implementing-a-"real-world"-scenario)
```
So 2.x becomes 3.x. Although I do appreciate real world scenario's I feel introducing SAP is a bit too much. Most people will probably not be aware of what this is so this might be too much cognitive load. What about using the using Blob and Table bindings in the activities and perhaps use a 3rd party API that doesn't require a registration and is centered around a basic concept that is well understood by everyone (such as the homework example).

lechnerc77 replied:
I was a bit afraid of that ... but kind of expected. Here is a proposal for an alternative scenario making use of the GitHub API (no authentication needed, no API key needed at least up to a certain limit):
- Story: We want to search for a repo and get some information about the author
- Step1 (Activity 1): Call the repo search API of GitHub and extract the author from the result https://api.github.com//search/repositories/{query}
- Step 2 (Activity 2):  Call the users API to extract further data about the author 
https://api.github.com/users/{userID}

Would that new story be okay for you? If yes I will make the corresponding changes.

In addition  I would like to put the advanced scenario in an kind of "outlook section" or appendix as I want to drag at least some SAP folks attention to Azure functions :-)

marcduiker replied:
Yes the GitHub use case sounds good. 

What about an additional homework section with the existing SAP part in it? We can explain in the homework section this is a sort of stretch goal for the homework and we can still mention it in the recording.

lechnerc77 replied:
Additional homework section sounds fine to me. Then _this is the way_

lechnerc77 replied:
Done

marcduiker commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,631 @@
+|2.3| [Implementation of the Activity Function `GetPgText`](#23-implementation-of-the-activity-function-`getpgtext`)
+|2.4| [Implementation of the Orchestrator Function](#24-implementation-of-the-orchestrator-function)
+|2.5| [Test the Implementation](#25-test-the-implementation)
+|3| [Retries - Dealing with Temporal Errors](#3-retries---dealing-with-temporal-errors)
```
So 3 becomes 4. Would need to be updated to match code in the previous exercise but in general this pattern is great to  include 👍.

marcduiker commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,631 @@
+|2.4| [Implementation of the Orchestrator Function](#24-implementation-of-the-orchestrator-function)
+|2.5| [Test the Implementation](#25-test-the-implementation)
+|3| [Retries - Dealing with Temporal Errors](#3-retries---dealing-with-temporal-errors)
+|4| [Racing Conditions - Dealing with Timeouts](#4-racing-conditions---dealing-with-timeouts)
```
4 becomes 5. Good idea to introduce the durable timer! I would rename this to Dealing with timeouts with a circuit breaker. Because racing conditions has a different meaning (more related to threading etc). Perhaps we can even link to this pattern: https://docs.microsoft.com/en-us/azure/architecture/patterns/circuit-breaker.

lechnerc77 replied:
Done

marcduiker commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,631 @@
+
+### 0.4 Prerequisites for local development
+
+As we could see Azure Durable Functions need some persistency attached. If you want to do a pure local development you must install the [Microsoft Azure Storage Emulator](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-emulator). Up to now the Durable Functions are not compatible with Azurite. So if you are on a non-Windows machine you must use a hybrid approach and connect your ...
```
Like I mentioned earlier, just a table with the prerequisites per exercise would be enough. There's already a prerequisites.md where you can link to that contains descriptions for these tools. If you think that file needs to be updated feel free to include that in this PR.

lechnerc77 replied:
Done

marcduiker commented on homework/durablefunctions/chaining/durable_locationsearch_api_ts.md:
```diff
@@ -0,0 +1,31 @@
+# Durable Functions -  Chaining Lesson Homework
```
Yes, really like this home work assignment! 👍

marcduiker commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,631 @@
+|2.5| [Test the Implementation](#25-test-the-implementation)
+|3| [Retries - Dealing with Temporal Errors](#3-retries---dealing-with-temporal-errors)
+|4| [Racing Conditions - Dealing with Timeouts](#4-racing-conditions---dealing-with-timeouts)
+|5| [Homework](#5-homework)
```
Home work and More info sections are good as they are, thanks! 👍
We need to figure out the previous and next links but we'll do that last moment.

lechnerc77 replied:
Done

marcduiker reviewed (changes requested):
First review finally done! 😅 I Didn't look into the technical details yet. Just focused on the story we're trying to tell. I really want to keep it as entry level friendly as possible. We can always come back to Durable Functions and incorporate more elaborate examples. As you've seen the timelines are shifted back a bit so no rush.  This content will go live early Feb by the looks of it (after the CosmosDB lesson).

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,631 @@
+|0.1| [Functions and Chaining](#01-functions-and-chaining)
+|0.2| [Solution via Durable Functions](#02-solution-via-durable-functions)
+|0.3| [Mechanics of Durable Functions](#03-mechanics-of-durable-functions)
+|0.4| [Prerequisites for local development](#04-prerequisites-for-local-development)
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,631 @@
+|0.2| [Solution via Durable Functions](#02-solution-via-durable-functions)
+|0.3| [Mechanics of Durable Functions](#03-mechanics-of-durable-functions)
+|0.4| [Prerequisites for local development](#04-prerequisites-for-local-development)
+|1| [Creating a Function App project for a Durable Function](#1-creating-a-function-app-project-for-a-durable-function)
```
Done

lechnerc77 commented:
@marcduiker : The review comments are done. Especially the restructuring of the lesson and the content (GitHub example) and the corresponding update of the homework section

marcduiker commented on homework/durablefunctions/chaining/durable_sapapi_ts.md:
```diff
@@ -0,0 +1,405 @@
+
+The goal of this homework is the development of a durable function setup to interact with an SAP backend system. 
+Assume the following scenario:
+Out function gets triggered via a purchase order create event that gives us an identifier for the order item. We want to identify the material group texts used in the order item. To do we must call a down stream system. We cannot retrieve the text in one call but we first make a call to get the material group ID and with this result issue a second call in order to retrieve the desired texts. This ...
```
```suggestion
Our function gets triggered via a purchase order create event that gives us an identifier for the order item. We want to identify the material group texts used in the order item. To do so we must call a down stream system. We cannot retrieve the text in one call but we first make a call to get the material group ID and with this result issue a second call in order to retrieve the desired texts. This represents a perfect scenario for a sequence of function calls. 

```

lechnerc77 replied:
Done + reformatted

marcduiker commented on homework/durablefunctions/chaining/durable_sapapi_ts.md:
```diff
@@ -0,0 +1,405 @@
+
+For this example we make use of a sandbox SAP system and the OData APIs available from there. To make the calls to the system a bit more convenient we make use of the [SAP Cloud SDK](https://community.sap.com/topics/cloud-sdk) that wraps the business objects in a virtual data model (VDM) and provides a fluent API to interact with the business object.
+
+The exercise i. e. the Activity Functions interact with an external API. This API is free to use and mimics an [SAP system](https://api.sap.com/package/SAPS4HANACloud?section=Artifacts). The only prerequisite is to register in order to get an API key. To get the free API keys to the sandbox system you must register yourself at the [API Business Hub](https://api.sap.com/) using the Log On button.
```
```suggestion
The exercise i.e. the Activity Functions interact with an external API. This API is free to use and mimics an [SAP system](https://api.sap.com/package/SAPS4HANACloud?section=Artifacts). The only prerequisite is to register in order to get an API key. To get the free API keys to the sandbox system you must register yourself at the [API Business Hub](https://api.sap.com/) using the Log On button.

```

lechnerc77 replied:
Done

marcduiker commented on homework/durablefunctions/chaining/durable_sapapi_ts.md:
```diff
@@ -0,0 +1,405 @@
+
+## Assignment
+
+Create a HTTP-triggered Durable Function consisting of two activities that need to be executed in sequence. The input parameters are 
```
```suggestion
Create an HTTP-triggered Durable Function consisting of two activities that need to be executed in sequence. The input parameters are:

```

lechnerc77 replied:
Done

marcduiker commented on homework/durablefunctions/chaining/durable_sapapi_ts.md:
```diff
@@ -0,0 +1,405 @@
+
+## A2 Retries - Dealing with Temporal Errors
+
+Next we implement the retry functionality. We want to separate the 
```
Sentence not complete?

lechnerc77 replied:
Rewrote sentence - done

marcduiker reviewed (changes requested):
We're nearly there! Most of the suggestions are small typo fixes and some additional explanation / restructuring of sentences. 

Given the fact that more languages will follow, can you move your source code files to be in `src\typescript\durablefunctions\` ? I will move the existing .NET code.

lechnerc77 commented on homework/durablefunctions/chaining/durable_sapapi_ts.md:
```diff
@@ -0,0 +1,405 @@
+
+## Goal 🎯
+
+The goal of this homework is the development of a durable function setup to interact with an SAP backend system. 
```
Done

lechnerc77 commented on homework/durablefunctions/chaining/durable_locationsearch_api_ts.md:
```diff
@@ -0,0 +1,31 @@
+
+## Share
+
+Please share you solutions on LinkedIn and Twitter using the #AzureFunctionsUniversity hashtag and mention us. We would love to see it!
```
Done

lechnerc77 commented on homework/durablefunctions/chaining/durable_locationsearch_api_ts.md:
```diff
@@ -0,0 +1,31 @@
+
+## Assignment
+
+Create a HTTP-triggered Durable Function consisting of two activities that need to be executed in sequence. The input is an URLencoded address (e. g. Seattle,%20Pioneer%20Square) where you want to do the local search and the [type of business entity](https://docs.microsoft.com/en-us/bingmaps/rest-services/common-parameters-and-types/type-identifiers/)(e.g. coffee) you want to search for.
```
Done

lechnerc77 commented on homework/durablefunctions/chaining/durable_locationsearch_api_ts.md:
```diff
@@ -0,0 +1,31 @@
+
+## Goal 🎯
+
+The goal for this lesson is to create a Durable Function that executes a local search for business entities around a certain address and returns the result. The function comprises two steps:
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+* Alternative to code-based workflows in Microsoft Azure: [Azure Logic Apps](https://azure.microsoft.com/en-us/services/logic-apps/)
+
+---
+[◀ Previous lesson](<previous>.md) | [🔼 Index](_index.md) | [Next lesson ▶](<next>.md)
```
done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+# Azure Durable Functions - Introduction & Chaining
```
Done

lechnerc77 commented on homework/durablefunctions/chaining/durable_locationsearch_api_ts.md:
```diff
@@ -0,0 +1,31 @@
+# Durable Functions -  Chaining Lesson Homework
```
Done, also changed the additional homework header accordingly

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+## 1. Introduction to Azure Durable Functions
+
+Within this section we want to take a look at the motivation for the usage of Azure Durable functions and take a look at the basic mechanics.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+### 1.1 Functions and Chaining
+
+In general, Functions are a great way to develop functionality in a serverless manner. However, this development should follow some best practices to avoid drawbacks or even errors when using them. The three main points to follow are:
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+* Functions should not call other functions
+* Functions should only do one thing well
+
+While the first practice is due to the nature of functions, the other two are technically possible but will contradicting the paradigms of serverless. In real life scenarios we often have to model processes that resemble a workflow, so we want to implement a sequence of single steps. How can we do that sticking to the best practices? One common solution for that is depicted below:
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+![Function Chaining Pattern](../../img/lessons/durablefunctions/chaining/functionchaining.png)
+
+Every function in the picture represents a single step of a workflow. In order to glue the functions together we use queues and storages or databases. So Function 1 is executed and stores its results in a table. Function 2 is triggered by an entry in the table via the corresponding bindings and gets executed representing the second step in the workflow. This story continued ten for Function 3. Th ...
```
done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+Every function in the picture represents a single step of a workflow. In order to glue the functions together we use queues and storages or databases. So Function 1 is executed and stores its results in a table. Function 2 is triggered by an entry in the table via the corresponding bindings and gets executed representing the second step in the workflow. This story continued ten for Function 3. Th ...
+
+* The single functions are just coupled via the event that they react to. From the outside it is not clear how the functions relate to each other although the represent a sequence of steps in a workflow.
```
done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+Every function in the picture represents a single step of a workflow. In order to glue the functions together we use queues and storages or databases. So Function 1 is executed and stores its results in a table. Function 2 is triggered by an entry in the table via the corresponding bindings and gets executed representing the second step in the workflow. This story continued ten for Function 3. Th ...
+
+* The single functions are just coupled via the event that they react to. From the outside it is not clear how the functions relate to each other although the represent a sequence of steps in a workflow.
+* The storage or queues in between single function executions are a necessary evil. One motivation for developing serverless is to care about servers less. Here we must care about technical infrastructure in order to cover the requirement.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+### 1.2 Solution via Durable Functions
+
+Azure Durable Functions are an extension to the Azure Functions Framework that support you with modeling workflows via Azure Functions. The extension supports you with dealing with all the tedious tasks mentioned above i. e. it does the heavy lifting for you, so that you can focus on the business requirement at hand. The local state is preserved by making use of [Event Sourcing](https://martinfow ...
```
done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+### 1.3 Mechanics of Durable Functions
+
+Durable Functions consist of three components:
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+![Durable Function Execution Schema](../../img/lessons/durablefunctions/chaining/SchemaDurableFunction0.png)
+
+The Client Function is triggered by an HTTP request and consequently triggers the Orchestrator function. Internally this means that a message is enqueued in a task hub. We do not have to care about that as we will se later.
```
Done + Capital "F" in Orchestrator Function

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+![Durable Function Execution Trigger](../../img/lessons/durablefunctions/chaining/SchemaDurableFunction1.png)
+
+After that the Client Functions scales back down to zero and the Orchestrator Function takes over. It fetches the task from the task hub and schedules the Activity Function.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+![Durable Function Execution Orchestrator](../../img/lessons/durablefunctions/chaining/SchemaDurableFunction2.png)
+
+Again the orchestrator scales down to zero and the Activity Function is executed and returns its results.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+![Durable Function Execution Activity](../../img/lessons/durablefunctions/chaining/SchemaDurableFunction3.png)
+
+After the execution of the first Activity Function the Orchestrator function is invoked again and checks if there are tasks left to do. In our scenario the second Activity Functions is executed. This cycle continuos until all Activity Function calls implemented in the Orchestrator are executed.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+After the execution of the first Activity Function the Orchestrator function is invoked again and checks if there are tasks left to do. In our scenario the second Activity Functions is executed. This cycle continuos until all Activity Function calls implemented in the Orchestrator are executed.
+
+After this more theoretical discussion Let us make our hands dirty with some code. But before let us sort our some prerequisites.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+## 2. Creating a Function App project for a Durable Function
+
+Ou scenario comprises a Durable Function App with two Activity Functions. The app will be triggered via an HTTP call.
```
Done.

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+## 2. Creating a Function App project for a Durable Function
+
+Ou scenario comprises a Durable Function App with two Activity Functions. The app will be triggered via an HTTP call.
+From a business perspective we get an purchase order item number handed over via the HTTP call. Based on the purchase order number we determine ina first step the material group of the purchase order item and in the consequent call we fetch the corresponding description in English.
```
Corrected the copy&paste error from the previous version.

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+### 2.2 The Orchestrator Function
+
+Now we create the Orchestrator Function responsible for the orchestration of the single Activity Functions.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+#### Steps
+
+1. Start the Azure Storage Emulator.
+2. Set the value of the `AzureWebJobsStorage` in your `local.settings.json` to `UseDevelopmentStorage=true`. This advices the function worker to use your local Azure Storage Emulator.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+4. Trigger the orchestration via the tool of your choice e.g. Postman.
+   > ❔ **Question** - What route do you have to use?
+
+   > 🔎 **Observation** - The result of the orchestration is not directly returned to the caller. Instead the Client Function is returning the ID and several HTTP endpoints to interact with the orchestration.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+## Goal 🎯
+
+The goal of this lesson is to give you an introduction into Azure Durable Functions including a first durable function that chains two functions calls. In addition we will take a look into some features of Durable Functions that help you when working with durable functions.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+|Nr|Exercise
+|-|-
+|0| [Prerequisites](#0-prerequisites)
+|1| [Introduction to Azure Durable Functions](#1-introduction-to-azure-durable-functions)
```
Makes sense. Adjusted the TOC

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+## 3. Implementing a "Real-World" Scenario
+
+In this section we develop some more realistic setup for our durable function. Assume the following situation:
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+#### Steps
+
+1. Import the oktokit core package.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   import { Octokit } from "@octokit/core"
+   ```
+
+2. Create an instance of the Oktokit class and build the query for the search based on the repository name.
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   return exactMatch.owner.login
+   ```
+
+The resulting function finally looks like this:
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+In this section we implement the Orchestrator Function that defines the call sequence of the Activity Functions and assures the transfer of the result of the first Activity Function to the second one.
+#### Steps
+
+1. Adopt the Orchestrator Function `GitHubInfoOrchestrator` to the Activity Functions. The first Activity Function is `GetRepositoryDetailsByName`. The returned information of the user ID must be transferred to the context of the Durable Function. The second Activity Function is `GetUserDetailsById`. The returned information of the second Activity Function is returned as response by the Orchestra ...
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+#### Steps
+
+1. Import the oktokit core package.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   import { Octokit } from "@octokit/core"
+   ```
+
+2. Create an instance of the Oktokit class and build the path for the request based on the user ID.
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   return userData
+   ```
+
+The resulting function finally looks like this:
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+In this section we implement the Orchestrator Function that defines the call sequence of the Activity Functions and assures the transfer of the result of the first Activity Function to the second one.
+#### Steps
+
+1. Adopt the Orchestrator Function `GitHubInfoOrchestrator` to the Activity Functions. The first Activity Function is `GetRepositoryDetailsByName`. The returned information of the user ID must be transferred to the context of the Durable Function. The second Activity Function is `GetUserDetailsById`. The returned information of the second Activity Function is returned as response by the Orchestra ...
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   import { Octokit } from "@octokit/core"
+   ```
+
+2. Create an instance of the Oktokit class and build the path for the request based on the user ID.
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   const searchResult = await octokit.request( apiPath )
+   ```
+
+4. As we only want the user data and no further information we restrict the result to this data and transform it into JSON. Be aware that you also nee to change the Promise type of the function to JSON.
````
done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   return userData
+   ```
+
+The resulting function finally looks like this:
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+## 4. Retries - Dealing with Temporal Errors
+
+As we are dealing with external systems the question is not if something will go wrong, but when this will be the case. So in this section we want to harden our setup to deal with temporal outages of the downstream system using retries when calling activities.
```
done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+## 4. Retries - Dealing with Temporal Errors
+
+As we are dealing with external systems the question is not if something will go wrong, but when this will be the case. So in this section we want to harden our setup to deal with temporal outages of the downstream system using retries when calling activities.
+Azure Durable functions have the built-in capability to execute an [automatic retry](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-error-handling?tabs=javascript#automatic-retry-on-failure) in case an Activity Function fails. The retry mechanism can be configured using a so called retry policy. In this section we will adopt our `GitHubInfoOrchestrator` to call t ...
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+   > 🔎 **Observation** - The constructor of the retry options calls only expects a minimum configuration.
+
+5. Adopt the calls of the Activity Functions to take into account retries and consider the retry configuration.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   const userInfos = yield context.df.callActivityWithRetry("GetUserDetailsById", retryConfig, context.bindingData.input)
+   ```
+
+6. In order to enforce an error we will cheat a bit. When we call the function we will hand over a parameter called `raiseException` that will be evaluated in the activity function `GetRepositoryDetailsByName`. In case that teh parameter is set to true we will throw an error. Change the logic of the activity function accordingly
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   ...
+   ```
+
+6. Start the durable function and make a call with correct input parameters to make sure we did not brake anything.
````
done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+   > 📝 **Tip** < TIP > - We make use of the replay mechanics of the durable function that is stored in the context. As soon as the execution is retried the input parameter gets corrected.
+
+9. Start the function and make a call enforcing the error.
```
done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+## 5. Circuit Breaker - Dealing with Timeouts
+
+In this section we want to become even more resilient with respect to the called system. In this section we want to deal with the scenario that the system will not return any response in a meaningful time. As a consequence we want to abort the orchestration if a certain time threshold es exceeded. To achieve this we introduce a time out.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   npm install moment
+   ```
+
+3. Set the  value of `"esModuleInterop"` to `true` in the `tsconfig.json` file. Otherwise the build of the project will fail
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+3. Set the  value of `"esModuleInterop"` to `true` in the `tsconfig.json` file. Otherwise the build of the project will fail
+
+4. Open the orchestrator function and import the moment package.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   const timeoutTask = context.df.createTimer(deadline.toDate())
+   ```
+
+   > 📝 **Tip** - Up to now we directly yielded all calls of the Activity Function. In this scenario we build up the activities and then let the durable function runtime execute them in parallel
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+   const repositoryDetailsTask = context.df.callActivity("GetRepositoryDetailsByName", context.bindingData.input)
+   ```
+
+7. Instruct the durable function runtime to let the two tasks (Activity Function and timer) race against each other. The durable function runtime will return the task that finishes first as the `winner`.
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+    }
+   ```
+
+   > 📝 **Tip** - The durable function runtime will not cancel any task but keep them running. Make sure that the loser task is canceled.
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+9. Start the durable function and make a call with correct input parameters to make sure we did not brake anything.
+
+   > ❔ **Question** - Check the execution in the storage explorer. What is different with respect to the original execution without the timer?
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
````diff
@@ -0,0 +1,633 @@
+     await sleep(10000)
+     ```
+
+11. Start the durable function and make a call with correct input parameters.
````
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+   > 📝 **Tip** - The durable function runtime will not cancel any task but keep them running. Make sure that the loser task is canceled.
+
+9. Start the durable function and make a call with correct input parameters to make sure we did not brake anything.
```
Added it for the real life scenario. I omitted this in the Hello World sceanrio as here I asked the URL as a question

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+### 3.1 Basic Setup of "Real-World" Scenario
+
+In this section we add the skeleton for the implementation.
+#### Steps
```
Done. 
Concerning the Client: I create the App in the same folder as the Hello World scenario, so I reuse the Client Function. I added a sentence in section 3.1 to highlight this procedure

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+   ![Binding for Orchestrator starter](../../img/lessons/durablefunctions/chaining/ClienFunctionConfig2.png)
+
+   > ❔ **Question** - Why is there an error in the `index.ts` file of the function?
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+   1. Select `Durable Functions orchestrator` as a template.
+   2. Name the function `DurableFunctionsOrchestrator`.
+
+   > 📝 **Tip** - Remove the comments from the index.ts file.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+   1. Select `Durable Functions activity` as a template.
+   2. Name the function `HelloCity`.
+
+   > 📝 **Tip** - Remove the comments from the index.ts file.
```
Done

lechnerc77 commented on lessons/durablefunctions/chaining-ts.md:
```diff
@@ -0,0 +1,633 @@
+
+#### Steps
+
+1. Import the oktokit core package.
```
Done

lechnerc77 commented:
> We're nearly there! Most of the suggestions are small typo fixes and some additional explanation / restructuring of sentences.
> 
> Given the fact that more languages will follow, can you move your source code files to be in `src\typescript\durablefunctions\` ? I will move the existing .NET code.

I moved the source code for the lesson and updated the paths pointing to it.
To stay consistent I also moved the source for the homework to this folder structure placing it below `src\typescript\durablefunctions\chaining\`

marcduiker reviewed (approved):
Thanks for all the changes and restructuring! Looking forward to the lesson live stream later this month!
